### DISCUSSION TITLE
Updating yabai (from latest HEAD) - Script

### DISCUSSION THREAD

**Discussion by gldtn** [asmvik/yabai#1904] [Show and tell]
Hello, just wanted to share an update script I put together..

I know there is a auto update in the wiki, but that has stopped working for me and every time I needed to update, I had to do every step manually.

With this script you can run an zsh alias, enter your sudo password and your done!

### **Change log**

_Nov. 17, 2023_
- fixed YABAI_CERT variable
- refactored script & added more comments

_Nov. 22, 2023_
- Added link to update script from my .dotfiles

### **Summary**

- pin/unpin Yabai to Brew
- export yabai-cert, re-installs yabai, codesigns it
- updates scripting edition

### **Observations:**

- This script pins/unpin yabai to brew to keep it from updating involuntarily when you upgrade you other packages.
- **You may need to change DIRs and YABAI_CERT name to match how you initially installed Yabai in your system.**

### **Let's get started**

**Create a file called update.sh with you favorite editor and save it to you Yabai config folder:**
_If you prefer you can download it from my .dotfiles here_ > [update.sh](https://github.com/gldtn/dotfiles/blob/main/.config/yabai/update.sh)
```bash
code $HOME/.config/yabai/update.sh
```

**Paste this code to it:**
_I'm assuming you kept the default name of YABAI_CERT=yabai-cert, else you'll need to change it._
```zsh
#!/bin/zsh

# scripting-addition;
# function to update sudoers file
function suyabai () {
    SHA256=$(shasum -a 256 $(which yabai) | awk "{print \$1;}")
    if [ -f "/private/etc/sudoers.d/yabai" ]; then
        sudo sed -i '' -e 's/sha256:[[:alnum:]]*/sha256:'${SHA256}'/' /private/etc/sudoers.d/yabai
        echo "sudoers > yabai > sha256 hash update complete"
    else
        echo "sudoers file does not exist yet. Please create one before running this script."
    fi
}

# check & unpin yabai from brew
if brew list --pinned | grep -q yabai; then
    brew unpin yabai
fi

# set cert & stop yabai services
export YABAI_CERT=yabai-cert
echo "Stopping yabai.."
yabai --stop-service

# reinstall yabai & codesign
echo "Updating yabai.."
brew reinstall koekeishiya/formulae/yabai
codesign -fs "${YABAI_CERT:-yabai-cert}" "$(brew --prefix yabai)/bin/yabai"

# update sudoers file & start yabai
suyabai
echo "Starting yabai.."
yabai --start-service

# pin yabai back to brew
brew pin yabai
if brew list --pinned | grep -q yabai; then
    echo "Yabai pinned to brew"
fi

# Success message
sleep 1
YABAI_V=$(yabai --version)
echo "Your running $YABAI_V"
echo "Yabai update completed successfully."
```

**Make the file executable:**
```bash
chmod +x $HOME/.config/yabai/update.sh
```

**Than you can execute it with:**
```bash
$HOME/.config/yabai/./update.sh
```

**Or you can add an alias to your .zshrc file like so:**
```bash
alias yabai.up='$HOME/.config/yabai/./update.sh'
```

Call it whatever you'd like by changing `yabai.up` and just call it in you fav terminal 😮‍💨

**Note**; _for Yabai stable release installs, use the modified version of this script put together by_ @tjex (https://github.com/koekeishiya/yabai/discussions/1904#discussioncomment-7720864)
> https://github.com/tjex/dotfiles/blob/mac/.local/scripts/sys-admin/update-yabai

### **Credits:**

- The `suyabai` function was borrowed from someone's .dotfiles which I don't remember who, so I can't give proper credit! 
- If someone with more expertise would like to chime in, please feel free! I have recently been updating this way without any errors.

**Comment by liuduwei** [comment]
thanks to share

**Comment by gldtn** [comment]
I updated the script to be a bit more friendly and check if yabai is currently pinned before trying to unpin it.

**Comment by rayhanadev** [comment]
Sweet!

**Comment by qhsong** [comment]
https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(from-HEAD)
This documents show how to   build code from real HEAD.  
It can fix may bugs in Macos Sonoma 14.1.1

**Comment by tjex** [comment]
Amazing! Thanks for the script. 
For those using the stable release (instead of HEAD), here is @gldtn script modified for our use case.
Essentially just removing the code signing and certificate env variable.

https://github.com/tjex/dotfiles/blob/mac/.local/scripts/sys-admin/update-yabai

**Reply by gldtn** [reply]
Thanks @tjex , I've added a link to your modified version to the bottom of the first post! Glad the script was helpful for you.
